---
title: "Figure S6: Functional annotation of genus pangenome"
author: "Payal Joglekar"
date: "`r Sys.Date()`"
output: html_document
---

###Background:
The purpose of this of this document is generate Figure S6 using the genus_pangenome matrix: genus_gene_presence_absence.csv. This file was generated using a reciprocal-best-blast approach for clustering cross-species orthologs from individual species pan-genomes. Note that Figure S6A was generated by manual counting of hypothetical genes based on the term "hypothetical" or with no annotation in both, prokka and emapper annotations for each gene. 

Steps:

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE, warning=FALSE, message=FALSE)
```

1. Load libraries

```{r cars}
library(tidyverse)
library(reshape2)
library(data.table)
library(ggVennDiagram)
library(RColorBrewer)
library(pheatmap)
library(stringr)
```

```{r, include=FALSE}
path <- getwd()
```

2. Figure S6A (percent hypothetical genes by gene-category (core/shell/cloud) as defined in Figure 2A)

```{r}
category <- c("core", "shell", "cloud")
annotated <- c(1546, 886, 2439)
annotatedp <- c(93, 79, 49)
unannotated <- c(114, 230, 2529)
unannotatedp <- c(7, 21, 51)
total <- c(1660, 1116, 4968)
totalp <- c(100, 100, 100)


a <- data.frame(category, annotatedp, unannotatedp)
a1 <- a %>% pivot_longer(cols=c(2,3), 
                         names_to = 'func', 
                         values_to = 'percent')


a1$category <- factor(a1$category, levels = c("core", "shell", "cloud"))
a1$func <- factor(a1$func, levels = c("unannotatedp", "annotatedp"))

col <- c(annotatedp = "coral3", unannotatedp = "gray")

ggplot(a1, aes(x = as.factor(category), y = percent)) +
  geom_col(aes(fill = as.factor(func)), position = "stack") +
  scale_fill_manual(values = col) + 
  scale_y_continuous(limits = c(0, 100))

#ggsave("gene_annotated_genus_pangenome.pdf", path = path, dpi = 300, width = 4, height = 4)

```


3. Figure S6B (pfam annotation by gene category)

3.1 Upload files and tidy them

```{r}
#pananroo gene_presence_absence
df <- fread(file.path(path,"genus_gene_presence_absence.csv"), data.table = FALSE) %>% as_tibble()

df1 <- column_to_rownames(df, var = "ID")
df1 <- df1 %>% select(3:128)

#metadata
meta <- read.csv(file.path(path,"metadata.csv"), header=T)
 
#trim metadata
meta <- meta %>% dplyr::rename(subject_ID = "subject.number.for.manuscript")
meta <- meta %>% select(c(isolate, genome_trim, species, subject_ID, site))

#change genome_trim names to isolate names
df1t <- as.data.frame(t(df1))
df1t <- rownames_to_column(df1t, var = "genome_trim")
meta1 <- meta %>% select(isolate, genome_trim)
df1t1 <- left_join(meta1, df1t, by = 'genome_trim')
df1t1 <- df1t1 %>% select(!genome_trim)
df1t1 <- column_to_rownames(df1t1, var = "isolate")
```

3.2 Calculate genes in each gene category (core/shell/cloud)
```{r}
#create a new column with # of genomes carrying a gene
df2 <- as.data.frame(t(df1t1)) #transpose
df2 <- df2 %>% mutate(sum = rowSums(df2)) #this gives number of genomes, temporary column
df2$sum <- as.numeric(df2$sum)

#all
dfall <- df2 
dfall <- dfall %>% filter(sum > 0)
dfall <- rownames_to_column(dfall, var = "gene")

#core to all three species (genus-core)
dfcore1 <- dfall %>% filter(sum >= 120)
dfcore <- df %>% filter(ID %in% dfcore1$gene) 

#genus_shell
dfshell1 <- dfall %>% filter(sum >= 19 & sum < 120)
dfshell <- df %>% filter(ID %in% dfshell1$gene) 

#genus-cloud
dfcloud1 <- dfall %>% filter(sum < 19)
dfcloud <- df %>% filter(ID %in% dfcloud1$gene) 
```


3.3 Pfam by gene category
```{r}
#pfam
#genus_core
dfcore_pfam <- dfcore %>% select(ID, SC_PFAMs, SE_PFAMs, SH_PFAMs)
#replace empty with NA
dfcore_pfam[dfcore_pfam == c("-")] <- NA 
dfcore_pfam[dfcore_pfam == c("?")] <- NA 
#coalesce all cog columns in one 
dfcore_pfam <- dfcore_pfam %>% mutate(core_pfam = coalesce(dfcore_pfam$SC_PFAMs, dfcore_pfam$SE_PFAMs, dfcore_pfam$SH_PFAMs))
#compute frequency of each pfam
dfcore_pfam_frq <- as.data.frame(table(dfcore_pfam$core_pfam))
dfcore_pfam_frq <- dfcore_pfam_frq %>% mutate(core_pfam = Freq/sum(Freq)*100)
dfcore_pfam_frq <- dfcore_pfam_frq %>% dplyr::rename(pfam = Var1)
dfcore_pfam_frq <- dfcore_pfam_frq %>% dplyr::rename(cor_pfam = Freq)
dfcore_pfam_frq1 <- dfcore_pfam_frq %>%                                      # Top N highest values by group
  arrange(desc(core_pfam)) %>% 
  dplyr::slice(1:20)
############################################################

#pfam
#genus_shell
dfshell_pfam <- dfshell %>% select(ID, SC_PFAMs, SE_PFAMs, SH_PFAMs)
#replace empty with NA
dfshell_pfam[dfshell_pfam == c("-")] <- NA 
dfshell_pfam[dfshell_pfam == c("?")] <- NA 
#coalesce all cog columns in one 
dfshell_pfam <- dfshell_pfam %>% mutate(shell_pfam = coalesce(dfshell_pfam$SC_PFAMs, dfshell_pfam$SE_PFAMs, dfshell_pfam$SH_PFAMs))
#compute frequency of each pfam
dfshell_pfam_frq <- as.data.frame(table(dfshell_pfam$shell_pfam))
dfshell_pfam_frq <- dfshell_pfam_frq %>% mutate(shell_pfam = Freq/sum(Freq)*100)
dfshell_pfam_frq <- dfshell_pfam_frq %>% dplyr::rename(pfam = Var1)
dfshell_pfam_frq <- dfshell_pfam_frq %>% dplyr::rename(s_pfam = Freq)
dfshell_pfam_frq1 <- dfshell_pfam_frq %>%                                      # Top N highest values by group
  arrange(desc(s_pfam)) %>% 
  dplyr::slice(1:20)
############################################################

#pfam
#genus_cloud
dfcloud_pfam <- dfcloud %>% select(ID, SC_PFAMs, SE_PFAMs, SH_PFAMs)
#replace empty with NA
dfcloud_pfam[dfcloud_pfam == c("-")] <- NA 
dfcloud_pfam[dfcloud_pfam == c("?")] <- NA 
#coalesce all cog columns in one 
dfcloud_pfam <- dfcloud_pfam %>% mutate(cloud_pfam = coalesce(dfcloud_pfam$SC_PFAMs, dfcloud_pfam$SE_PFAMs, dfcloud_pfam$SH_PFAMs))
#compute frequency of each pfam
dfcloud_pfam_frq <- as.data.frame(table(dfcloud_pfam$cloud_pfam))
dfcloud_pfam_frq <- dfcloud_pfam_frq %>% mutate(cloud_pfam = Freq/sum(Freq)*100)
dfcloud_pfam_frq <- dfcloud_pfam_frq %>% dplyr::rename(pfam = Var1)
dfcloud_pfam_frq <- dfcloud_pfam_frq %>% dplyr::rename(cl_pfam = Freq)
dfcloud_pfam_frq1 <- dfcloud_pfam_frq %>%                                      # Top N highest values by group
  arrange(desc(cl_pfam)) %>% 
  dplyr::slice(1:20)
############################################################

#join both lists
pfam1 <- left_join(dfcore_pfam_frq1, dfshell_pfam_frq, by = "pfam") %>% 
  left_join(dfcloud_pfam_frq,by = "pfam")

pfam2 <- left_join(dfshell_pfam_frq1, dfcore_pfam_frq, by = "pfam") %>% 
  left_join(dfcloud_pfam_frq,by = "pfam")
  
pfam3 <- left_join(dfcloud_pfam_frq1, dfshell_pfam_frq, by = "pfam") %>% 
  left_join(dfcore_pfam_frq,by = "pfam")
  
#put all data frames into list
pfam_list <- list(pfam1, pfam2, pfam3)

#merge all data frames in list
pfam <- pfam_list %>% purrr::reduce(full_join, by=c('pfam', "cor_pfam", "s_pfam", "cl_pfam", "core_pfam", "shell_pfam", "cloud_pfam"))
pfam[is.na(pfam)] <- 0
#write_csv(pfam, "genus_pangenome_staph_Pfam_core_accessory.csv")

############################################################
#plot
p <- pfam %>% select(1,2,4,6)
p <- p %>% dplyr::rename(c(core = cor_pfam, shell = s_pfam, cloud = cl_pfam))
p1 <- p %>% pivot_longer(cols = (c(2:4)),
                                 names_to = "genome",
                                 values_to = "abundance")


p1$pfam <- gsub("Arm-DNA-bind_4,Phage_int_SAM_3,Phage_integrase", "Phage_integrase", p1$pfam)

p1$genome <- factor(p1$genome, levels = c("cloud", "shell", "core"))

p1$pfam <- factor(p1$pfam, levels = c("ABC_tran", "BPD_transp_1", "SBP_bac_5", "ABC_membrane,ABC_tran", "ABC2_membrane_3", "FecCD", "MFS_1", "AA_permease", "AA_permease_2", "Response_reg,Trans_reg_C", "adh_short_C2", "Acetyltransf_1", "Acetyltransf_3", "Aldedh", "ADH_N,ADH_zinc_N", "M20_dimer,Peptidase_M20", "2-Hacid_dh,2-Hacid_dh_C", "GHMP_kinases_C,GHMP_kinases_N", "Pribosyltran", "Thioredoxin", "Hydrolase_3", "Hydrolase_4", "AA_kinase", "NUDIX", "TetR_N", "MarR", "HTH_1,LysR_substrate", "ArsB", "Abi", "Staph_haemo", "DUF4064", "DUF1474", "DUF1433", "HTH_7,Resolvase", "DDE_Tnp_IS240", "DUF772", "Rep_1", "DDE_Tnp_1,DUF772", "HTH_28", "DUF600", "Peptidase_M78", "CHAP", "Phage_integrase", "HTH_3", "HTH_26", "Terminase_2", "DnaB_2", "DUF3310", "Sipho_tail", "Methylase_S", "N6_N4_Mtase", "DNA_methylase", "N6_Mtase", "HNH"))


#bubbleplot
col <- c("core" = "#993902", "shell" = "#fdb82c", "cloud" = "#f5f500") #"#ffff5f"
p2 <- ggplot(data=p1, mapping = aes(x = as.factor(pfam), y = as.factor(genome), size = abundance)) + geom_point(aes(fill = genome), colour="black",pch=21) +
  scale_fill_manual(values = col) +
  scale_size_area(max_size = 6) + theme(axis.text.x = element_text(size= 5, angle=90, hjust = 1))
p2

#ggsave(p2, file="genus_pfam2.pdf", width=6, height=2.2, dpi = 300)
```

4 Figure S6C (kegg enrichment by gene category)

4.1 Generate KO lists from emapper KEGG annotations
```{r}
#Steps for Microbiomeprofiler
#############KEGG ENRICHMENT using eggnog mapper###################
#make  a list of all kegg ko numbers from the eggNOG output file
all_kegg <- df %>% select(ID, SC_KEGG_ko, SE_KEGG_ko, SH_KEGG_ko)

#replace empty with NA
all_kegg[all_kegg == c("-")] <- NA 
all_kegg[all_kegg == c("?")] <- NA 
#coalesce all cog columns in one 
all_kegg <- all_kegg %>% mutate(all_kegg = coalesce(all_kegg$SC_KEGG_ko, all_kegg$SE_KEGG_ko, all_kegg$SH_KEGG_ko))

gene2ko <- all_kegg %>%
  dplyr::select(Gene = ID, KEGG_ko = all_kegg) %>%
  na.omit()

#clean up by removing the "ko:" in front of every KO term
gene2ko[,2]<-gsub("ko:","",gene2ko[,2])
head(gene2ko)

# expand, since some genes/proteins will have multiple assigned KO terms
gene2ko <- data.table(gene2ko)
gene2ko <- gene2ko[, list(KEGG_ko = unlist(strsplit(KEGG_ko , ","))), by = Gene]

write.csv(gene2ko, file = file.path(path, "genus_pangenome_staph_genus_all_keggmap.csv"))

#########################################################################
#make  a list of all kegg ko numbers in different gene categories
#core
df1_kegg <-dfcore %>% select(ID, SC_KEGG_ko, SE_KEGG_ko, SH_KEGG_ko)

#replace empty with NA
df1_kegg[df1_kegg == c("-")] <- NA 
df1_kegg[df1_kegg == c("?")] <- NA 
#coalesce all cog columns in one 
df1_kegg <- df1_kegg %>% mutate(kegg = coalesce(df1_kegg$SC_KEGG_ko, df1_kegg$SE_KEGG_ko, df1_kegg$SH_KEGG_ko))

gene2ko1 <- df1_kegg %>%
  dplyr::select(Gene = ID, KEGG_ko = kegg) %>%
  na.omit()

#clean up by removing the "ko:" in front of every KO term
gene2ko1[,2]<-gsub("ko:","",gene2ko1[,2])
head(gene2ko1)

# expand, since some genes/proteins will have multiple assigned KO terms
gene2ko1 <- data.table(gene2ko1)
gene2ko1 <- gene2ko1[, list(KEGG_ko = unlist(strsplit(KEGG_ko , ","))), by = Gene]

write.csv(gene2ko1, file = file.path(path, "genus_pangenome_staph_genus_core_keggmap.csv"))

#######################################################
#shell

df2_kegg <-dfshell %>% select(ID, SC_KEGG_ko, SE_KEGG_ko, SH_KEGG_ko)

#replace empty with NA
df2_kegg[df2_kegg == c("-")] <- NA 
df2_kegg[df2_kegg == c("?")] <- NA 
#coalesce all cog columns in one 
df2_kegg <- df2_kegg %>% mutate(kegg = coalesce(df2_kegg$SC_KEGG_ko, df2_kegg$SE_KEGG_ko, df2_kegg$SH_KEGG_ko))

gene2ko2 <- df2_kegg %>%
  dplyr::select(Gene = ID, KEGG_ko = kegg) %>%
  na.omit()

#clean up by removing the "ko:" in front of every KO term
gene2ko2[,2]<-gsub("ko:","",gene2ko2[,2])
head(gene2ko2)

# expand, since some genes/proteins will have multiple assigned KO terms
gene2ko2 <- data.table(gene2ko2)
gene2ko2 <- gene2ko2[, list(KEGG_ko = unlist(strsplit(KEGG_ko , ","))), by = Gene]

write.csv(gene2ko2, file = file.path(path, "genus_pangenome_staph_shell_keggmap.csv"))


#######################################################
#cloud
df3_kegg <-dfcloud %>% select(ID, SC_KEGG_ko, SE_KEGG_ko, SH_KEGG_ko)

#replace empty with NA
df3_kegg[df3_kegg == c("-")] <- NA 
df3_kegg[df3_kegg == c("?")] <- NA 
#coalesce all cog columns in one 
df3_kegg <- df3_kegg %>% mutate(kegg = coalesce(df3_kegg$SC_KEGG_ko, df3_kegg$SE_KEGG_ko, df3_kegg$SH_KEGG_ko))

gene2ko3 <- df3_kegg %>%
  dplyr::select(Gene = ID, KEGG_ko = kegg) %>%
  na.omit()

#clean up by removing the "ko:" in front of every KO term
gene2ko3[,2]<-gsub("ko:","",gene2ko3[,2])
head(gene2ko3)

# expand, since some genes/proteins will have multiple assigned KO terms
gene2ko3 <- data.table(gene2ko3)
gene2ko3 <- gene2ko3[, list(KEGG_ko = unlist(strsplit(KEGG_ko , ","))), by = Gene]

write.csv(gene2ko3, file = file.path(path, "genus_pangenome_staph_cloud_keggmap.csv"))
#######################################################
```


4.2 Use the saved KO lists in MicrobiomeProfiler to find KEGG gene enrichment

```{r}
############MicrobiomeProfiler##########################
#BiocManager::install("MicrobiomeProfiler", force = TRUE)
library(MicrobiomeProfiler)
run_MicrobiomeProfiler()
```

4.3 Plot KEGG enrichemnt

```{r}
#Using MicrobiomeProfiler outputs from each category generate a combined "KEGG_enrichment_pangenome_MicrobiomeProfiler.csv" file
#Columns: "KEGG Pathway Description", "Gene category", 	"GeneRatio", 	"p_adjst", 	"Genecount"

#plot 
kegg_enrich <- read.csv(file.path(path,"KEGG_enrichment_pangenome_MicrobiomeProfiler.csv")) #created manually from MicrobiomeProfiler output


#set factors for gene categogry
kegg_enrich$category <- factor(kegg_enrich$category, levels = c("core", "shell", "cloud"))

#plot
kc <- ggplot(kegg_enrich, aes(x = GeneRatio, y = reorder(Description, GeneRatio), fill = p_adjst)) + geom_col(width = 0.8) + geom_text(aes(label = Genecount), hjust = 1, color = "white") + facet_grid(as.factor(category)~., scales = "free", space = "free") 
kc
ggsave("genus_pangenome_kegg_enrichment.pdf", path = path, dpi = 300, width = 8, height = 4)
```